// jscpd/build/listeners/statistic.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var events_1 = require("../events");
var jscpd_1 = require("../jscpd");
var models_1 = require("../stores/models");
var options_1 = require("../utils/options");
var StatisticListener = /** @class */ (function () {
    function StatisticListener(options) {
        this.options = options;
        this.statistic = {
            detectionDate: new Date().toISOString(),
            formats: {},
            total: StatisticListener.getDefaultStatistic()
        };
        this.statistic.threshold = this.options.threshold;
    }
    StatisticListener.getDefaultStatistic = function () {
        return {
            lines: 0,
            sources: 0,
            clones: 0,
            duplicatedLines: 0,
            percentage: 0,
            newDuplicatedLines: 0,
            newClones: 0
        };
    };
    StatisticListener.prototype.attach = function (eventEmitter) {
        eventEmitter.on(events_1.MATCH_SOURCE_EVENT, this.matchSource.bind(this));
        eventEmitter.on(events_1.CLONE_FOUND_EVENT, this.cloneFound.bind(this));
    };
    StatisticListener.prototype.cloneFound = function (clone) {
        var id = clone.duplicationA.sourceId;
        var id2 = clone.duplicationB.sourceId;
        var linesCount = clone.duplicationA.end.line - clone.duplicationA.start.line;
        this.statistic.total.clones++;
        this.statistic.total.duplicatedLines += linesCount;
        this.statistic.formats[clone.format].total.clones++;
        this.statistic.formats[clone.format].total.duplicatedLines += linesCount;
        this.statistic.formats[clone.format].sources[id].clones++;
        this.statistic.formats[clone.format].sources[id].duplicatedLines += linesCount;
        this.statistic.formats[clone.format].sources[id2].clones++;
        this.statistic.formats[clone.format].sources[id2].duplicatedLines += linesCount;
        this.updatePercentage(clone.format);
        this.saveStatistic();
    };
    StatisticListener.prototype.matchSource = function (source) {
        source.format = source.format || 'javascript';
        if (!this.statistic.formats.hasOwnProperty(source.format)) {
            this.statistic.formats[source.format] = {
                sources: {},
                total: StatisticListener.getDefaultStatistic()
            };
        }
        this.statistic.total.sources++;
        this.statistic.total.lines += source.lines;
        this.statistic.formats[source.format].total.sources++;
        this.statistic.formats[source.format].total.lines += source.lines;
        this.statistic.formats[source.format].sources[source.id] =
            this.statistic.formats[source.format].sources[source.id] || StatisticListener.getDefaultStatistic();
        this.statistic.formats[source.format].sources[source.id].sources = 1;
        this.statistic.formats[source.format].sources[source.id].lines += source.lines;
        this.updatePercentage(source.format);
        this.saveStatistic();
    };
    StatisticListener.prototype.saveStatistic = function () {
        var statisticStore = jscpd_1.getStoreManager().getStore(models_1.STATISTIC_DB);
        statisticStore.set(options_1.getOption('executionId', this.options), this.statistic);
    };
    StatisticListener.prototype.updatePercentage = function (format) {
        var _this = this;
        this.statistic.total.percentage = this.calculatePercentage(this.statistic.total.lines, this.statistic.total.duplicatedLines);
        this.statistic.formats[format].total.percentage = this.calculatePercentage(this.statistic.formats[format].total.lines, this.statistic.formats[format].total.duplicatedLines);
        Object.entries(this.statistic.formats[format].sources).forEach(function (_a) {
            var _b = __read(_a, 2), id = _b[0], stat = _b[1];
            _this.statistic.formats[format].sources[id].percentage = _this.calculatePercentage(stat.lines, stat.duplicatedLines);
        });
    };
    StatisticListener.prototype.calculatePercentage = function (totalLines, clonedLines) {
        return totalLines ? Math.round((10000 * clonedLines) / totalLines) / 100 : 0.0;
    };
    return StatisticListener;
}());
exports.StatisticListener = StatisticListener;
