// jscpd/build/tokenizer/prism.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var prismjs_1 = require("prismjs");
var formats_1 = require("./formats/formats");
var loadLanguages = require('prismjs/components/');
var ignore = {
    ignore: [
        {
            pattern: /(jscpd:ignore-start)[\s\S]*?(?=jscpd:ignore-end)/,
            lookbehind: true,
            greedy: true
        },
        {
            pattern: /jscpd:ignore-start/,
            greedy: false
        },
        {
            pattern: /jscpd:ignore-end/,
            greedy: false
        }
    ]
};
var punctuation = {
    new_line: /\n/,
    empty: /\s+/
};
prismjs_1.languages.markup.script.inside = __assign({}, ignore, prismjs_1.languages.markup.script.inside, punctuation);
prismjs_1.languages.markup.style.inside = __assign({}, ignore, prismjs_1.languages.markup.style.inside, punctuation);
function getLanguagePrismName(lang) {
    if (formats_1.FORMATS.hasOwnProperty(lang) && formats_1.FORMATS[lang].parent) {
        return formats_1.FORMATS[lang].parent;
    }
    return lang;
}
function initLanguages(langs) {
    loadLanguages(langs.map(getLanguagePrismName));
    Object.keys(prismjs_1.languages).forEach(function (lang) {
        prismjs_1.languages[lang] =
            typeof prismjs_1.languages[lang] === 'object' ? __assign({}, ignore, prismjs_1.languages[lang], punctuation) : prismjs_1.languages[lang];
    });
}
exports.initLanguages = initLanguages;
function tokenize(code, language) {
    var length = 0;
    var line = 1;
    var column = 1;
    initLanguages([language]);
    var tokens = [];
    prismjs_1.tokenize(code, prismjs_1.languages[getLanguagePrismName(language)]).forEach(function (t) { return (tokens = tokens.concat(createTokens(t, language))); });
    function sanitizeLangName(name) {
        return name && name.replace ? name.replace('language-', '') : 'unknown';
    }
    function createTokenFromString(token, lang) {
        return [
            {
                format: lang,
                type: 'default',
                value: token,
                length: token.length
            }
        ];
    }
    function createTokenFromFlatToken(token, lang) {
        return [
            {
                format: lang,
                type: token.type,
                value: token.content,
                length: token.length
            }
        ];
    }
    function createTokens(token, lang) {
        if (token instanceof prismjs_1.Token && typeof token.content === 'string') {
            return createTokenFromFlatToken(token, lang);
        }
        if (token instanceof prismjs_1.Token && Array.isArray(token.content)) {
            var res_1 = [];
            token.content.forEach(function (t) { return (res_1 = res_1.concat(createTokens(t, token.alias ? sanitizeLangName(token.alias) : lang))); });
            return res_1;
        }
        return createTokenFromString(token, lang);
    }
    function calculateLocation(token) {
        var result = token;
        var lines = result.value.split('\n');
        var newLines = lines.length - 1;
        var start = {
            line: line,
            column: column
        };
        column = newLines !== 0 ? lines[lines.length - 1].length + 1 : column + lines[lines.length - 1].length;
        var end = {
            line: line + newLines,
            column: column
        };
        result.loc = { start: start, end: end };
        result.range = [length, length + result.length];
        length += result.length;
        line += newLines;
        return result;
    }
    return tokens.map(calculateLocation).filter(function (t) {
        return t.format !== 'important' && t.format !== 'property' && t.format !== 'url';
    });
}
exports.tokenize = tokenize;
