// jscpd/build/tokenizer/token-map.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("../utils");
var TOKEN_VALUE_HASH_LENGTH = 10;
var TOKEN_TYPE_HASH_LENGTH = 10;
var TOKEN_HASH_LENGTH = TOKEN_VALUE_HASH_LENGTH + TOKEN_TYPE_HASH_LENGTH;
function createTokenHash(token) {
    return utils_1.md5(token.type).substr(0, TOKEN_TYPE_HASH_LENGTH) + utils_1.md5(token.value).substr(0, TOKEN_VALUE_HASH_LENGTH);
}
var TokensMap = /** @class */ (function () {
    function TokensMap(tokens, format, minTokens) {
        this.tokens = tokens;
        this.format = format;
        this.minTokens = minTokens;
        this.position = 0;
        this.map = this.tokens.map(function (token) { return createTokenHash(token); }).join('');
    }
    TokensMap.prototype.getStartPosition = function () {
        return this.tokens[0].range[0];
    };
    TokensMap.prototype.getEndPosition = function () {
        return this.tokens[this.getLength() - 1].range[1];
    };
    TokensMap.prototype.getFormat = function () {
        return this.format;
    };
    TokensMap.prototype.getLength = function () {
        return this.tokens.length;
    };
    TokensMap.prototype.getSourceId = function () {
        return this.sourceId;
    };
    TokensMap.prototype.setSourceId = function (id) {
        this.sourceId = id;
    };
    TokensMap.prototype[Symbol.iterator] = function () {
        return this;
    };
    TokensMap.prototype.next = function () {
        var result;
        var mapFrame = utils_1.md5(this.map.substring(this.position * TOKEN_HASH_LENGTH, this.position * TOKEN_HASH_LENGTH + this.minTokens * TOKEN_HASH_LENGTH)).substring(0, 10);
        if (this.position < this.tokens.length - this.minTokens) {
            result = {
                done: false,
                value: {
                    id: mapFrame,
                    format: this.getFormat(),
                    sourceId: this.getSourceId(),
                    start: this.tokens[this.position],
                    end: this.tokens[this.position + this.minTokens]
                }
            };
            this.position++;
        }
        else {
            result = {
                done: true,
                value: {}
            };
        }
        return result;
    };
    return TokensMap;
}());
exports.TokensMap = TokensMap;
