// jscpd/build/clone/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var models_1 = require("../stores/models");
var stores_manager_1 = require("../stores/stores-manager");
var source_1 = require("../utils/source");
function createClone(startMap, endMap) {
    return __awaiter(this, void 0, void 0, function () {
        var format, hashesStore, sourceStart, sourceEnd;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    format = startMap.format;
                    hashesStore = stores_manager_1.StoresManager.getStore(models_1.getHashDbName(format));
                    return [4 /*yield*/, hashesStore.get(startMap.id)];
                case 1:
                    sourceStart = _a.sent();
                    return [4 /*yield*/, hashesStore.get(endMap.id)];
                case 2:
                    sourceEnd = _a.sent();
                    return [2 /*return*/, {
                            format: format,
                            foundDate: new Date().getTime(),
                            duplicationA: {
                                sourceId: startMap.sourceId,
                                start: startMap.start.loc.start,
                                end: endMap.end.loc.end,
                                range: [startMap.start.range[0], endMap.end.range[1]],
                                fragment: getFragment(startMap.sourceId, startMap.start.range[0], endMap.end.range[1])
                            },
                            duplicationB: {
                                sourceId: sourceStart.sourceId,
                                start: sourceStart.start.loc.start,
                                end: sourceEnd.end.loc.end,
                                range: [sourceStart.start.range[0], sourceEnd.end.range[1]],
                                fragment: getFragment(sourceStart.sourceId, sourceStart.start.range[0], sourceEnd.end.range[1])
                            }
                        }];
            }
        });
    });
}
exports.createClone = createClone;
function getFragment(id, start, end) {
    return source_1.sourceToString({ id: id }).substring(start, end);
}
exports.getFragment = getFragment;
function getSourceFragment(source, start, end) {
    return source_1.sourceToString(source).substring(start, end);
}
exports.getSourceFragment = getSourceFragment;
function getSourceFragmentLength(source, start, end) {
    return getSourceFragment(source, start, end).split('\n').length;
}
exports.getSourceFragmentLength = getSourceFragmentLength;
function isCloneLinesBiggerLimit(clone, minLines) {
    return clone.duplicationA.end.line - clone.duplicationA.start.line >= minLines;
}
exports.isCloneLinesBiggerLimit = isCloneLinesBiggerLimit;
