// jscpd/build/reporters/console.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var cli_table3_1 = __importDefault(require("cli-table3"));
var safe_1 = require("colors/safe");
var events_1 = require("../events");
var utils_1 = require("../utils");
var ConsoleReporter = /** @class */ (function () {
    function ConsoleReporter(options) {
        this.options = options;
    }
    ConsoleReporter.prototype.attach = function (eventEmitter) {
        eventEmitter.on(events_1.CLONE_FOUND_EVENT, this.cloneFound.bind(this));
    };
    ConsoleReporter.prototype.report = function () {
        var _this = this;
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        var _a = __read(args, 2), statistic = _a[1];
        if (statistic) {
            var table_1 = new cli_table3_1.default({
                head: ['Format', 'Files analyzed', 'Total lines', 'Clones found', 'Duplicated lines', '%']
            });
            Object.keys(statistic.formats)
                .filter(function (format) { return statistic.formats[format].sources; })
                .forEach(function (format) {
                table_1.push(_this.convertStatisticToArray(format, statistic.formats[format].total));
            });
            table_1.push(this.convertStatisticToArray(safe_1.bold('Total:'), statistic.total));
            console.log(table_1.toString());
        }
    };
    ConsoleReporter.prototype.cloneFound = function (clone) {
        var duplicationA = clone.duplicationA, duplicationB = clone.duplicationB, format = clone.format;
        console.log('Clone found (' + format + '):' + (clone.isNew ? safe_1.red('*') : ''));
        console.log(" - " + utils_1.getPathConsoleString(this.options, duplicationA.sourceId) + " [" + utils_1.getSourceLocation(duplicationA.start, duplicationA.end) + "]");
        console.log("   " + utils_1.getPathConsoleString(this.options, duplicationB.sourceId) + " [" + utils_1.getSourceLocation(duplicationB.start, duplicationB.end) + "]");
        console.log('');
    };
    ConsoleReporter.prototype.convertStatisticToArray = function (format, statistic) {
        return [
            format,
            "" + statistic.sources,
            "" + statistic.lines,
            "" + statistic.clones,
            "" + statistic.duplicatedLines,
            statistic.percentage + "%"
        ];
    };
    return ConsoleReporter;
}());
exports.ConsoleReporter = ConsoleReporter;
