// jscpd/build/reporters/console-full.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var bytes = require("bytes");
var cli_table3_1 = __importDefault(require("cli-table3"));
var safe_1 = require("colors/safe");
var events_1 = require("../events");
var utils_1 = require("../utils");
var console_1 = require("./console");
var TABLE_OPTIONS = {
    chars: {
        top: '',
        'top-mid': '',
        'top-left': '',
        'top-right': '',
        bottom: '',
        'bottom-mid': '',
        'bottom-left': '',
        'bottom-right': '',
        left: '',
        'left-mid': '',
        mid: '',
        'mid-mid': '',
        right: '',
        'right-mid': '',
        middle: '│'
    }
};
var ConsoleFullReporter = /** @class */ (function (_super) {
    __extends(ConsoleFullReporter, _super);
    function ConsoleFullReporter(options) {
        return _super.call(this, options) || this;
    }
    ConsoleFullReporter.prototype.attach = function (eventEmitter) {
        eventEmitter.on(events_1.SOURCE_SKIPPED_EVENT, this.skipSource.bind(this));
    };
    ConsoleFullReporter.prototype.report = function (clones) {
        var _this = this;
        clones.forEach(function (clone) {
            _this.cloneFullFound(clone);
        });
    };
    ConsoleFullReporter.prototype.skipSource = function (source) {
        console.log(safe_1.grey("Skipped " + source.path + " (Size: " + bytes(source.size) + (source.lines ? ', Lines: ' + source.lines : '') + ")"));
    };
    ConsoleFullReporter.prototype.cloneFullFound = function (clone) {
        if (this.options.reporters && this.options.reporters.includes('consoleFull')) {
            var table_1 = new cli_table3_1.default(TABLE_OPTIONS);
            this.cloneFound(clone);
            clone.duplicationA.fragment.split('\n').forEach(function (line, position) {
                table_1.push(utils_1.generateLine(clone, position, line));
            });
            console.log(table_1.toString());
            console.log('');
        }
    };
    return ConsoleFullReporter;
}(console_1.ConsoleReporter));
exports.ConsoleFullReporter = ConsoleFullReporter;
