// jscpd/build/reporters/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var use_1 = require("../utils/use");
var console_1 = require("./console");
var console_full_1 = require("./console-full");
var exec_timer_1 = require("./exec-timer");
var html_1 = require("./html");
var json_1 = require("./json");
var silent_1 = require("./silent");
var threshold_1 = require("./threshold");
var time_1 = require("./time");
var verbose_1 = require("./verbose");
var xml_1 = require("./xml");
var EXISTING_REPORTERS = {
    console: console_1.ConsoleReporter,
    consoleFull: console_full_1.ConsoleFullReporter,
    time: time_1.TimeReporter,
    json: json_1.JsonReporter,
    xml: xml_1.XmlReporter,
    html: html_1.HtmlReporter,
    silent: silent_1.SilentReporter,
    threshold: threshold_1.ThresholdReporter,
    verbose: verbose_1.VerboseReporter,
    execTimer: exec_timer_1.ExecTimerReporter
};
var REPORTERS = {};
function registerReporter(name, reporter) {
    REPORTERS[name] = reporter;
}
exports.registerReporter = registerReporter;
function getRegisteredReporters() {
    return REPORTERS;
}
exports.getRegisteredReporters = getRegisteredReporters;
function registerReportersByName(options) {
    var _a = options.reporters, reporters = _a === void 0 ? [] : _a;
    reporters.forEach(function (rep) {
        var reporter = EXISTING_REPORTERS[rep] || use_1.useReporter(rep);
        registerReporter(rep, new reporter(options));
    });
}
exports.registerReportersByName = registerReportersByName;
