// jscpd/build/reporters/json.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fs_1 = require("fs");
var fs_extra_1 = require("fs-extra");
var utils_1 = require("../utils");
var options_1 = require("../utils/options");
var JsonReporter = /** @class */ (function () {
    function JsonReporter(options) {
        this.options = options;
        this.json = {
            duplicates: [],
            statistics: {}
        };
    }
    JsonReporter.prototype.attach = function () { };
    JsonReporter.prototype.report = function (clones, statistic) {
        var _this = this;
        if (statistic) {
            this.json.statistics = statistic;
        }
        clones.forEach(function (clone) {
            _this.cloneFound(clone);
        });
        fs_extra_1.ensureDirSync(options_1.getOption('output', this.options));
        fs_1.writeFileSync(options_1.getOption('output', this.options) + '/jscpd-report.json', JSON.stringify(this.json, null, '\t'));
    };
    JsonReporter.prototype.cloneFound = function (clone) {
        var startLineA = clone.duplicationA.start.line;
        var endLineA = clone.duplicationA.end.line;
        var startLineB = clone.duplicationB.start.line;
        var endLineB = clone.duplicationB.end.line;
        this.json.duplicates.push({
            format: clone.format,
            lines: endLineA - startLineA + 1,
            fragment: clone.duplicationA.fragment,
            tokens: 0,
            firstFile: {
                name: utils_1.getPath(this.options, clone.duplicationA.sourceId),
                start: startLineA,
                end: endLineA,
                startLoc: clone.duplicationA.start,
                endLoc: clone.duplicationA.end,
                blame: clone.duplicationA.blame
            },
            secondFile: {
                name: utils_1.getPath(this.options, clone.duplicationB.sourceId),
                start: startLineB,
                end: endLineB,
                startLoc: clone.duplicationB.start,
                endLoc: clone.duplicationB.end,
                blame: clone.duplicationB.blame
            }
        });
    };
    return JsonReporter;
}());
exports.JsonReporter = JsonReporter;
