// jscpd/build/reporters/verbose.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var bytes = require("bytes");
var safe_1 = require("colors/safe");
var events_1 = require("../events");
var VerboseReporter = /** @class */ (function () {
    function VerboseReporter(options) {
        this.options = options;
        this.sourceCount = 0;
        this.startTime = process.hrtime();
    }
    VerboseReporter.prototype.attach = function (eventEmitter) {
        eventEmitter.on(events_1.MATCH_SOURCE_EVENT, this.matchSource.bind(this));
        eventEmitter.on(events_1.CLONE_FOUND_EVENT, this.cloneFound.bind(this));
        eventEmitter.on(events_1.SOURCE_SKIPPED_EVENT, this.skipSource.bind(this));
    };
    VerboseReporter.prototype.report = function () { };
    VerboseReporter.prototype.matchSource = function (source) {
        this.sourceCount++;
        console.log(safe_1.green('Source matched:'));
        console.log(safe_1.grey(JSON.stringify(source, null, '\t')));
        this.generateStatistic(source.format);
    };
    VerboseReporter.prototype.skipSource = function (source) {
        console.log(safe_1.grey("Source skipped " + source.path + " (Size: " + bytes(source.size) + (source.lines ? ', Lines: ' + source.lines : '') + ")"));
    };
    VerboseReporter.prototype.cloneFound = function (clone) {
        console.log(safe_1.red('Clone found:'));
        console.log(safe_1.grey(JSON.stringify(clone, null, '\t')));
        this.generateStatistic(clone.format);
    };
    VerboseReporter.prototype.generateStatistic = function (format) {
        console.log(safe_1.bgBlue([parseHrtimeToSeconds(process.hrtime(this.startTime)), 'sec', format, this.sourceCount].join(' ')));
    };
    return VerboseReporter;
}());
exports.VerboseReporter = VerboseReporter;
function parseHrtimeToSeconds(hrtime) {
    return (hrtime[0] + hrtime[1] / 1e9).toFixed(3);
}
