// jscpd/build/stores/leveldb.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs_extra_1 = require("fs-extra");
var rimraf_1 = __importDefault(require("rimraf"));
var level = require('level');
var LevelDbStore = /** @class */ (function () {
    function LevelDbStore(options) {
        this.options = options;
        if (!options.persist) {
            rimraf_1.default.sync(".jscpd/" + this.options.name);
        }
        fs_extra_1.ensureDirSync(".jscpd/" + this.options.name);
        this.db = level(".jscpd/" + this.options.name);
    }
    LevelDbStore.prototype.get = function (key) {
        return this.db
            .get(key)
            .then(function (value) { return JSON.parse(value); })
            .catch(function () { return undefined; });
    };
    LevelDbStore.prototype.getAllByKeys = function (keys) {
        var _this = this;
        return Promise.all(keys.map(function (i) { return _this.get(i); }));
    };
    LevelDbStore.prototype.set = function (key, value) {
        return this.db.put(key, JSON.stringify(value));
    };
    LevelDbStore.prototype.init = function () {
        return Promise.resolve({});
    };
    LevelDbStore.prototype.has = function (key) {
        return this.db
            .get(key)
            .then(function () { return true; })
            .catch(function () { return false; });
    };
    LevelDbStore.prototype.hasKeys = function (keys) {
        var _this = this;
        return Promise.all(keys.map(function (k) { return _this.has(k); }));
    };
    LevelDbStore.prototype.connect = function () {
        this.db.open();
        return Promise.resolve();
    };
    LevelDbStore.prototype.delete = function (key) {
        return this.db.del(key);
    };
    LevelDbStore.prototype.update = function (key, value) {
        var _this = this;
        return this.delete(key).then(function () { return _this.set(key, value); });
    };
    LevelDbStore.prototype.close = function () {
        var _this = this;
        return new Promise(function (resolve) {
            _this.db.close(function () {
                if (!_this.options.persist) {
                    rimraf_1.default(".jscpd/" + _this.options.name, { maxBusyTries: 10 }, function (err) {
                        if (err) {
                            console.log(err);
                        }
                        resolve();
                    });
                }
                else {
                    resolve();
                }
            });
        });
    };
    return LevelDbStore;
}());
exports.LevelDbStore = LevelDbStore;
