// jscpd/build/stores/memory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var MemoryStore = /** @class */ (function () {
    function MemoryStore() {
        this.values = {};
    }
    // constructor(private options: IStoreOptions) {
    //   if (this.options.name === 'files') {
    //     console.log(this.options);
    //   }
    // }
    MemoryStore.prototype.get = function (key) {
        return Promise.resolve(this.values[key]);
    };
    MemoryStore.prototype.getAllByKeys = function (keys) {
        var _this = this;
        return Promise.resolve(keys.map(function (key) { return _this.values[key]; }));
    };
    MemoryStore.prototype.set = function (key, value) {
        this.values[key] = value;
        return Promise.resolve(value);
    };
    MemoryStore.prototype.init = function (values) {
        this.values = values;
        return Promise.resolve(values);
    };
    MemoryStore.prototype.has = function (key) {
        return Promise.resolve(this.values.hasOwnProperty(key));
    };
    MemoryStore.prototype.hasKeys = function (keys) {
        var _this = this;
        return Promise.resolve(keys.map(function (key) { return _this.values.hasOwnProperty(key); }));
    };
    MemoryStore.prototype.connect = function () {
        return Promise.resolve();
    };
    MemoryStore.prototype.delete = function (key) {
        delete this.values[key];
        return Promise.resolve();
    };
    MemoryStore.prototype.update = function (key, value) {
        this.values[key] = value;
        return Promise.resolve(value);
    };
    MemoryStore.prototype.close = function () {
        return Promise.resolve();
    };
    return MemoryStore;
}());
exports.MemoryStore = MemoryStore;
