// jscpd/build/stores/stores-manager.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs_1 = require("fs");
var rimraf_1 = __importDefault(require("rimraf"));
var use_1 = require("../utils/use");
var files_1 = require("./files");
var leveldb_1 = require("./leveldb");
var memory_1 = require("./memory");
var StoreManager = /** @class */ (function () {
    function StoreManager() {
        this.registeredStores = {
            memory: memory_1.MemoryStore,
            files: files_1.FilesStore,
            leveldb: leveldb_1.LevelDbStore
        };
        this.stores = {};
        this.options = {};
    }
    StoreManager.prototype.initialize = function (options) {
        if (options === void 0) { options = {}; }
        this.options = options;
    };
    StoreManager.prototype.flush = function () {
        this.stores = {};
    };
    StoreManager.prototype.close = function () {
        var _this = this;
        return Promise.all(Object.values(this.stores).map(function (store) { return store.close(); })).then(function () {
            _this.flush();
            var path = '.jscpd';
            if (fs_1.existsSync(path)) {
                var subFolders = fs_1.readdirSync(path);
                if (subFolders.length === 0) {
                    rimraf_1.default(path, {}, function (err) {
                        if (err) {
                            console.log(err);
                        }
                    });
                }
            }
        });
    };
    StoreManager.prototype.getStore = function (name) {
        if (!this.has(name)) {
            this.create(name);
        }
        return this.stores[name];
    };
    StoreManager.prototype.has = function (name) {
        return this.stores.hasOwnProperty(name);
    };
    StoreManager.prototype.getRegisteredStore = function (type) {
        if (!this.isRegistered(type)) {
            this.registeredStores[type] = use_1.use(type, use_1.ModuleType.db);
        }
        return this.registeredStores[type];
    };
    StoreManager.prototype.isRegistered = function (type) {
        return this.registeredStores.hasOwnProperty(type);
    };
    StoreManager.prototype.create = function (name) {
        if (!this.has(name)) {
            // hashes.javascript
            var _a = __read(name.split('.'), 1), main = _a[0];
            var _b = this.options[name] ||
                this.options[main] ||
                this.options['*'] || { type: 'leveldb' }, type = _b.type, _c = _b.options, options = _c === void 0 ? {} : _c;
            this.stores[name] = new (this.getRegisteredStore(type))(__assign({}, options, { name: name }));
            this.stores[name].connect();
        }
    };
    return StoreManager;
}());
exports.StoreManager = StoreManager;
exports.StoresManager = new StoreManager();
