// jscpd/build/utils/options.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs_1 = require("fs");
var fs_extra_1 = require("fs-extra");
var path_1 = require("path");
var formats_1 = require("../tokenizer/formats");
function getOption(name, options) {
    return options ? options[name] || getDefaultOptions()[name] : getDefaultOptions()[name];
}
exports.getOption = getOption;
function prepareOptions(cli) {
    var config = cli.config ? path_1.resolve(cli.config) : path_1.resolve('.jscpd.json');
    var storedConfig = {};
    var argsConfig;
    var packageJsonConfig;
    argsConfig = {
        minTokens: cli.minTokens ? Number(cli.minTokens) : undefined,
        minLines: cli.minLines ? Number(cli.minLines) : undefined,
        maxLines: cli.maxLines ? Number(cli.maxLines) : undefined,
        maxSize: cli.maxSize,
        debug: cli.debug,
        executionId: cli.executionId,
        silent: cli.silent,
        blame: cli.blame,
        cache: cli.cache,
        output: cli.output,
        xslHref: cli.xslHref,
        format: cli.format,
        formatsExts: parseFormatsExtensions(cli.formatsExts),
        list: cli.list,
        mode: cli.mode,
        absolute: cli.absolute,
        noSymlinks: cli.noSymlinks,
        ignoreCase: cli.ignoreCase,
        gitignore: cli.gitignore
    };
    if (cli.threshold !== undefined) {
        argsConfig.threshold = Number(cli.threshold);
    }
    if (cli.reporters) {
        argsConfig.reporters = cli.reporters.split(',');
    }
    if (cli.format) {
        argsConfig.format = cli.format.split(',');
    }
    if (cli.ignore) {
        argsConfig.ignore = cli.ignore.split(',');
    }
    argsConfig.path = cli.path ? [cli.path].concat(cli.args) : cli.args;
    Object.keys(argsConfig).forEach(function (key) {
        if (typeof argsConfig[key] === 'undefined') {
            delete argsConfig[key];
        }
    });
    if (!fs_1.existsSync(config)) {
        config = '';
    }
    else {
        storedConfig = fs_extra_1.readJSONSync(config);
    }
    if (fs_1.existsSync(process.cwd() + '/package.json')) {
        packageJsonConfig = fs_extra_1.readJSONSync(process.cwd() + '/package.json').jscpd || {};
    }
    var result = __assign({ config: config }, getDefaultOptions(), packageJsonConfig, storedConfig, argsConfig);
    if (result.hasOwnProperty('config') && result.config && path_1.isAbsolute(result.config) && result.path) {
        result.path = result.path.map(function (path) { return path_1.resolve(path_1.dirname(config), path); });
    }
    result.reporters = result.reporters || [];
    result.listeners = result.listeners || [];
    if (result.silent) {
        result.reporters = result.reporters.filter(function (reporter) { return reporter.indexOf('console') === -1; }).concat('silent');
    }
    if (result.threshold !== undefined) {
        result.reporters = __spread(result.reporters, ['threshold']);
    }
    result.reporters = __spread(result.reporters, ['time']);
    result.reporters = __spread(new Set(result.reporters));
    return result;
}
exports.prepareOptions = prepareOptions;
function getDefaultOptions() {
    return {
        executionId: new Date().toISOString(),
        path: [process.cwd()],
        minLines: 5,
        maxLines: 500,
        maxSize: '30kb',
        minTokens: 50,
        output: './report',
        reporters: ['console', 'time'],
        listeners: ['statistic'],
        ignore: [],
        mode: 'mild',
        threshold: undefined,
        format: __spread(formats_1.getSupportedFormats()),
        formatsExts: {},
        debug: false,
        silent: false,
        blame: false,
        cache: true,
        absolute: false,
        noSymlinks: false,
        ignoreCase: false,
        gitignore: false,
        reportersOptions: {}
    };
}
exports.getDefaultOptions = getDefaultOptions;
function parseFormatsExtensions(extensions) {
    var result = {};
    if (extensions) {
        extensions.split(';').forEach(function (format) {
            var pair = format.split(':');
            result[pair[0]] = pair[1].split(',');
        });
    }
    else {
        return undefined;
    }
    return result;
}
