// jscpd/build/utils/use.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var detectInstalled = require('detect-installed');
var ModuleType;
(function (ModuleType) {
    ModuleType["reporter"] = "reporter";
    ModuleType["mode"] = "mode";
    ModuleType["db"] = "db";
    ModuleType["tokenizer"] = "tokenizer";
})(ModuleType = exports.ModuleType || (exports.ModuleType = {}));
/**
 * import reporter
 * @param name
 * @deprecated
 */
function useReporter(name) {
    return use(name, ModuleType.reporter);
}
exports.useReporter = useReporter;
/**
 * import mode
 * @param name
 * @deprecated
 */
function useMode(name) {
    return use(name, ModuleType.mode);
}
exports.useMode = useMode;
function use(name, type) {
    var packageName = "jscpd-" + name + "-" + type;
    if (!detectInstalled.sync(packageName, { local: true })) {
        throw new Error("Module (type: " + type + ") \"" + packageName + "\" does not found, please check that you have installed the package");
    }
    return require(packageName).default;
}
exports.use = use;
